gnomad_qc/v5/sample_qc/merge_project_meta.py

```diff
@@ -0,0 +1,371 @@
+    )
+    return ht.key_by(sample_id_field)
+
+
+def main():
+    """Create v5 project metadata."""
+    meta_hts = {}
+    for project, project_metadata in get_config().items():
+        project_hts = []
+        logger.info("Processing the %s project's metadata...", project)
+        for data_type in project_metadata.keys():
+            data_type_hts = []
+            logger.info("Importing %s's %s metadata ...", project, data_type)
+            for file in project_metadata[data_type]:
+
+                if file["file_format"] == "ht":
+                    ht = file["ht"]
+
+                elif file["file_format"] == "tsv":
+                    file_types = {
+                        input_field: FINAL_SCHEMA_FIELDS_AND_TYPES[final_field]
+                        for final_field, input_field in file["field_mappings"].items()
+                    }
+                    ht = hl.import_table(file["path"], types=file_types)
+
+                else:
+                    raise ValueError(f"File format {file['file_type']} not supported")
+
+                ht = ht.flatten().select_globals()
+                ht = ht.select(
+                    **{
+                        new_field: ht[field]
+                        for new_field, field in file["field_mappings"].items()
+                    }
+                )
+                ht = ht.key_by("s")
+                logger.info("Selecting out fields for the final schema...")
+                ht = select_only_final_fields(ht, project, data_type)
+                data_type_hts.append(ht)
+
+            if len(data_type_hts) > 1:
+                logger.info(
+                    "Combining the %s metadata files from %s...", data_type, project
+                )
+                data_type_ht = reduce(
+                    (lambda joined_ht, ht: joined_ht.join(ht, how="outer")),
+                    data_type_hts,
+                )
+            else:
+                data_type_ht = data_type_hts[0]
+
+            data_type_ht = data_type_ht.annotate(data_type=data_type)
+            logger.info("Updating the %s metadata to the final schema...", data_type)
+            data_type_ht = update_ht_to_final_schema(data_type_ht)
+            project_hts.append(data_type_ht.key_by("s"))
+
+        logger.info(
+            "Combining the %s project's files into a single metadata file...", project
+        )
+        project_ht = reduce((lambda joined_ht, ht: joined_ht.union(ht)), project_hts)
+        project_ht = project_ht.annotate(project=project)
+        logger.info("Updating the %s metadata to the final schema...", project)
+        project_ht = update_ht_to_final_schema(project_ht)
+
+        meta_hts[project] = project_ht
+
+    sample_collisions = get_sample_collisions(meta_hts)
+    logger.info(
+        f"Samples that appear in more than one project: {hl.eval(sample_collisions)}"
+    )
+
+    for project in get_config().keys():
+        meta_hts[project] = add_project_prefix_to_sample_ids(
```

Comment:
looks like this is rerunning things so maybe a checkpoint can be added?

Reply:
Hopefully the checkpoint above handles this